python/cudf/cudf/core/join/join.py

```diff
@@ -201,24 +209,28 @@ def perform_merge(self) -> Frame:
         )
 
         gather_index = self._using_left_index or self._using_right_index
+        lkwargs = {
+            "gather_map": left_rows,
+            "nullify": True,
+            "check_bounds": False,
+        }
+        rkwargs = {
+            "gather_map": right_rows,
+            "nullify": True,
+            "check_bounds": False,
+        }
+        if not self.lhs_is_index:
+            lkwargs["keep_index"] = gather_index
+        if not self.rhs_is_index:
+            rkwargs["keep_index"] = gather_index
```

Comment:
I'm not sure that I like storing all the kwargs in a dict just so that one of them can be modified this way, I find it less readable. Why not simply change the calls to `_gather` to `keep_index=True if self.lhs_is_index else gather_index`?

Reply:
The usual way to call is
```python
         left_result = (
            (self.lhs._gather(
                left_rows, nullify=True, check_bounds=False
                ) if self.lhs_is_index
                  else self.lhs._gather(
                    left_rows, nullify=True, check_bounds=False, keep_index=gather_index
                )
            )
            if left_rows is not None
            else cudf.core.frame.Frame()
        )
        right_result = (
            (self.rhs._gather(
                right_rows, nullify=True, check_bounds=False
                ) if self.rhs_is_index
                  else self.rhs._gather(
                    right_rows, nullify=True, check_bounds=False, keep_index=gather_index
                )
            )
            if right_rows is not None
            else cudf.core.frame.Frame()
        )
```

Reply:
Right, but if you don't pass a keyword argument it just uses the default parameter. My suggestion was just to explicitly pass the default parameter value in in the `[lr]hs_is_index` case.

Reply:
`_base_index._gather` no longer has `keep_index` (nor variadic kwargs). You can't explicitly specify it otherwise you will run into issue like this:
```python
>>> def f(): pass
>>> f(a=128)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: f() got an unexpected keyword argument '
```
I guess the crux is that `_base_index` and `IndexFrame` has no relationship so the signature of `_gather` is not bounded by any constraints.

Reply:
Ohhh I see the problem. OK, that makes sense. We'll have to see whether there's a better solution to this problem in the future... But this is fine for now, then.